cips/cip-tx-limits.md

```diff
@@ -0,0 +1,44 @@
+| cip | TBD(27) |
+| - | - |
+| title | Block limits for number of PFBs/non-PFBs and transaction sizes |
+| description | Set limits for number of PFBs and non-PFBs per block and transaction size |
+| author | Josh Stein ([@jcstein](https://github.com/jcstein)), Nina Barbakadze ([@ninabarbakadze](https://github.com/ninabarbakadze)) |
+| discussions-to |  <https://forum.celestia.org/t/cip-limit-number-of-pfbs-and-non-pfbs-per-block-increase-transaction-size-limit/1843> |
+| status | Draft |
+| type | Standards Track |
+| category | Core |
+| created | 2024-10-16 |
+
+## Abstract
+
+This CIP proposes to set limits for the number of PayForBlobs (PFBs) messages and non-PFBs messages per block and set the limit for transaction size. The proposal is to set the limits to 600 PFBs messages and 200 non-PFBs messages per block and set the transaction size limit to 2MiB. Setting PFB and non-PFBs limits is not consensus-breaking. Setting the transaction size limit is consensus-breaking.
+
+## Specification
+
+1. The number of PFBs per block is limited to 600 by setting `MaxPFBMessages` to 600. `MaxPFBMessages` is the maximum number of PFB messages a block can contain.
+
+1. The number of non-PFBs messages per block is limited to 200 by setting `MaxNonPFBMessages` to 200. `MaxNonPFBMessages` is the maximum number of SDK messages, aside from PFBs, that a block can contain.
+
+1. It's important to note that these limits are not strictly enforced. While they are defined by the `celestia-app` implementation, a validator could potentially modify the `PrepareProposal` logic, run a custom binary, and produce blocks that exceed the specified limits for PFB or non-PFBs transactions.
+
+1. The size of a transaction is limited to 2MiB by setting `MaxTxSize` to 2097152, which is 2MiB in bytes. From version v3 and above, in `CheckTx`, `PrepareProposal`, and `ProcessProposal`, each transaction's size is checked against the `appconsts.MaxTxSize` threshold. This ensures that transactions over the limit are rejected or excluded at all stages, from initial submission to execution.
+
+## Rationale
+
+The rationale for this proposal is twofold:
+
+1. To prevent long block times on the network by limiting the number of PFBs and non-PFBs messages per block. This was initially not considered consensus-breaking, but it has a meaningful effect on users and should be formalized in a CIP.
```

Comment:
We can do this at a later point but it might be helpful to remember later on how these numbers were chosen

Reply:
we could point to this PR https://github.com/celestiaorg/celestia-app/pull/390 as I am adding the results to it and will open it for review afterwards

Reply:
The PR 309 referenced is from deciding on the prefix for addresses and unrelated to this? https://github.com/celestiaorg/celestia-app/issues/390
<img width="349" alt="Screenshot 2024-10-17 at 12 26 35" src="https://github.com/user-attachments/assets/f1fe924e-f910-4c2b-b487-c54a55111b15">

@rach-id Can you please provide a condensed explanation of what you learned in the PR that was intended to be linked?

Reply:
my bad, this PR: https://github.com/celestiaorg/celestia-app/pull/3904

That PR contains benchmarks that we run to measure the processing time different ABCI methods take with different transaction types.

Then, to limit that processing time to ~0.25sec, we added a soft limiter in prepare proposal so that the default block created by the block construction mechanism respects that processing time limit. 

This soft limiter is a simple cap on the number of messages a block can contain, based on running the PRs benchmarks on 4 CPU 16GB RAM machines, which is the suggested validator configuration.

This way, the default blocks don't take much time to process. However, if a block not respecting those limits reaches consensus, it still can be included and won't be wasted.